gossip: Allow deferring advertise of local node

asias opened the pull request:
Currently the replacing node sets the status as STATUS_UNKNOWN when it
starts gossip service for the first time before it sets the status to
HIBERNATE to start the replacing operation. This introduces the
following race:

1) Replacing node using the same IP address of the node to be replaced
starts gossip service without setting the gossip STATUS (will be seen as
STATUS_UNKNOWN by other nodes)

2) Replacing node waits for gossip to settle and learns status and
tokens of existing nodes

3) Replacing node announces the HIBERNATE STATUS.

After Step 1 and before Step 3, existing nodes will mark the replacing
node as UP, but haven't marked the replacing node as doing replacing
yet. As a result, the replacing node will not be excluded from the read
replicas and will be considered a target node to serve CQL reads.

To fix, we start gossip service on the replacing node with a gossip generation
zero so that existing nodes will ignore the replacing node because
they have bigger generation number.

When the replacing node can advertise HIBERNATE status, it sets the generation
number to normal so other nodes can learn it from gossip.

Fixes #7312

tgrabiec commented:
This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes. 

`gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.

Not sending responses seems like a safer way to implement a shadow round.

asias commented:
> This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> 
> `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.

How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.

> 
> Not sending responses seems like a safer way to implement a shadow round.

With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode.  It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode. 

Btw, I plan to remove all gossiper::is_in_shadow_round() logic completely once all supported release understand the new GOSSIP_GET_ENDPOINT_STATES verb.

asias commented:
Changes in v2: 2433be57128bea72eac82e88bba08833eaccc7fc
- Avoid responding to echo message in case advertise_myself is false

avikivity commented:
@tgrabiec bump

tgrabiec commented on service/storage_service.cc:
```diff
@@ -930,6 +932,8 @@ void storage_service::bootstrap() {
             { gms::application_state::CDC_STREAMS_TIMESTAMP, versioned_value::cdc_streams_timestamp(_cdc_streams_ts) },
             { gms::application_state::STATUS, versioned_value::hibernate(true) },
         }).get();
+        sleep_abortable(std::chrono::seconds(30), _abort_source).get();
```
Why did you add this 30s wait?

asias replied:
It is a temporary debug patch sneaked in during rebase. I will drop it.

tgrabiec commented:
> > This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> > `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.
> 
> How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.

When the node restarts, it will indeed inject an entry into the endpoint_state_map (storage_service does this) with generation 0, which is equal to the "shadow" generation used by this patch, so "shadow" gossip will be accepted, not ignored as we want.

Secondly, it's not the gossiper component who injects those entries, but scylla's storage service. It's better if gossiper's `advertise_myself::no` is not coupled with external things like that because they can change independently of the gossiper. Suppressing messages within the gossiper doesn't rely on external things like this, so it's better.

> 
> > Not sending responses seems like a safer way to implement a shadow round.
> 
> With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode. It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode.

The `advertise_myself` mode is also a new mode which can cause regressions. I'm not sure why you think the long shadow round will be less safe. The long shadow round only suppresses messages. Other nodes will perceive it the same way as if the node was partitioned. Sending no messages seems safer than actually sending messages but with a special generation number, because it doesn't affect the state of other nodes.

asias commented:
> > > This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> > > `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.
> > 
> > 
> > How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.
> 
> When the node restarts, it will indeed inject an entry into the endpoint_state_map (storage_service does this) with generation 0, which is equal to the "shadow" generation used by this patch, so "shadow" gossip will be accepted, not ignored as we want.
> 
> Secondly, it's not the gossiper component who injects those entries, but scylla's storage service. It's better if gossiper's `advertise_myself::no` is not coupled with external things like that because they can change independently of the gossiper. Suppressing messages within the gossiper doesn't rely on external things like this, so it's better.


Good that you agree that the !es path is not possible as you claimed earlier. The node will run the remote_generation == local_generation path. However, we do not even need storage service to inject the zero generation to make it work. No node in a working state will have generation zero, such node should never be marked as alive by other nodes. We can skip to mark a node with zero generation to be alive. 


> 
> > > Not sending responses seems like a safer way to implement a shadow round.
> > 
> > 
> > With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode. It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode.
> 
> The `advertise_myself` mode is also a new mode which can cause regressions. I'm not sure why you think the long shadow round will be less safe. The long shadow round only suppresses messages. Other nodes will perceive it the same way as if the node was partitioned. Sending no messages seems safer than actually sending messages but with a special generation number, because it doesn't affect the state of other nodes.



 No, this patch introduces no new mode. The gossip protocol is not changed. Look at the gossip handlers, nothing has changed there in this patch.  No special handling for shadow round. This patch only changes the way the node choose the gossip generation. 

It is not sending any message in most of the cases because other nodes have bigger generation. In special cases when other nodes also think the node has a zero generation, this patch will still work even with this version, because echo message to this replacing node will fail. It is even more safe with v2 below, because other nodes will not try to mark a node with zero generation at all. 

More importantly, this patch fixes an real bug in your proposal - not sending message is not enough.

asias commented:
Changes in v2:
- Dropped a debug sleep introduced during rebase
- Added a patch to skip mark node with zero generation to be alive

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1717,20 +1722,24 @@ void gossiper::examine_gossiper(utils::chunked_vector<gossip_digest>& g_digest_l
     // Although gossiper runs on cpu0 only, we need to listen incoming gossip
     // message on all cpus and forard them to cpu0 to process.
-    return get_gossiper().invoke_on_all([do_bind] (gossiper& g) {
+    return get_gossiper().invoke_on_all([do_bind, advertise] (gossiper& g) {
```
capture of `advertise` is unused

asias replied:
Yes. Removed. Thanks.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1717,20 +1722,24 @@ void gossiper::examine_gossiper(utils::chunked_vector<gossip_digest>& g_digest_l
             logger.warn("Use the generation number provided by user: generation = {}", generation_nbr);
         }
+        if (!advertise) {
+            _initial_generation_number = generation_nbr;
```
This needs a comment explaining what's going on.

bhalevy replied:
Also, when are we using `force_gossip_generation`?
It feels dangerous to me in conjunction with this change, can we get rid of it?

bhalevy replied:
Should we assert that `generation_nbr != 0` here?

asias replied:
The force_gossip_generation is introduced (743b529c2b035658d33afd14f6a01beff3097b12) to help user fix cluster without patch 0a52ecb6df3deec729759ea87b4986334f207c07. We need some more time to get rid of this option.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1753,6 +1762,20 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
+    return make_ready_future<>();
+}
+
+
```
nit: extra blank line

asias replied:
Removed. Thanks.

bhalevy commented on gms/gossiper.hh:
```diff
@@ -87,6 +87,8 @@ class feature_service;
 struct bind_messaging_port_tag {};
 using bind_messaging_port = bool_class<bind_messaging_port_tag>;
 
+using advertise_myself_t = bool_class<class advertise_myself_tag>;
```
nit: don't need the `_t` suffix.
seems to go against the conventional use of `bool_class` elsewhere.
Even, `bind_messaging_port` right above.

tgrabiec replied:
> > > > This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> > > > `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.
> > > 
> > > 
> > > How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.
> > 
> > 
> > When the node restarts, it will indeed inject an entry into the endpoint_state_map (storage_service does this) with generation 0, which is equal to the "shadow" generation used by this patch, so "shadow" gossip will be accepted, not ignored as we want.
> > Secondly, it's not the gossiper component who injects those entries, but scylla's storage service. It's better if gossiper's `advertise_myself::no` is not coupled with external things like that because they can change independently of the gossiper. Suppressing messages within the gossiper doesn't rely on external things like this, so it's better.
> 
> Good that you agree that the !es path is not possible as you claimed earlier. The node will run the remote_generation == local_generation path. However, we do not even need storage service to inject the zero generation to make it work. No node in a working state will have generation zero, such node should never be marked as alive by other nodes. We can skip to mark a node with zero generation to be alive.

I see a contradiction now, and this highlights that there are actually two problems we try to solve here. The contradiction is this: If we're fine with overriding endpoint states on the other nodes, and we only want to prevent the replacing node to be marked alive, then we don't need to gossip with fake generation, we can gossip with the standard generation and only have the replacing node not respond to those requests when it's in the shadow mode.

If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.

I don't know if we want to prevent overriding the state. It seems safer to not override it, but I can't think of any problem with overriding either.

> 
> > > > Not sending responses seems like a safer way to implement a shadow round.
> > > 
> > > 
> > > With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode. It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode.
> > 
> > 
> > The `advertise_myself` mode is also a new mode which can cause regressions. I'm not sure why you think the long shadow round will be less safe. The long shadow round only suppresses messages. Other nodes will perceive it the same way as if the node was partitioned. Sending no messages seems safer than actually sending messages but with a special generation number, because it doesn't affect the state of other nodes.
> 
> No, this patch introduces no new mode. The gossip protocol is not changed. Look at the gossip handlers, nothing has changed there in this patch. No special handling for shadow round. This patch only changes the way the node choose the gossip generation.

Suppressing messages is not exactly changing the protocol, other nodes won't notice anything. Why do you dislike changing the handlers?

> 
> It is not sending any message in most of the cases because other nodes have bigger generation. In special cases when other nodes also think the node has a zero generation, this patch will still work even with this version, because echo message to this replacing node will fail. It is even more safe with v2 below, because other nodes will not try to mark a node with zero generation at all.
> 
> More importantly, this patch fixes an real bug in your proposal - not sending message is not enough.

Yes, but the same fix can be applied to my proposal as well, it doesn't invalidate it.

asias replied:
I used advertise_myself_t because I want to use advertise_myself as the variable name.

asias replied:
But I found a way to solve this and drop the _t suffix.

bhalevy replied:
You can still use it both as a type and the parameter name.  For example:
```
enum class yes_or_no { no, yes };

bool check(yes_or_no yes_or_no) {
    return yes_or_no == yes_or_no::yes;
}
```

bhalevy commented on gms/gossiper.cc:
```diff
@@ -409,6 +409,9 @@ future<> gossiper::handle_ack2_msg(gossip_digest_ack2 msg) {
 }
 
 future<> gossiper::handle_echo_msg() {
+    if (_initial_generation_number) {
```
This is a bit confusing since it holds an optional number so from reading it, without being aware it's optional, it looks like this expression is testing the value, rather the optional state.

I suggest calling `_initial_generation_number.has_value()` instead to make it clear.

asias replied:
It is a pretty standard way to use if (foo_opt) to check if the opt contains a value. I do not mind using has_value to make it clear.

bhalevy replied:
True.  The reason it is more confusing here, is that optional value is numeric and so the member name suggests.
Until I realized it holds an optional<uint32_t>, I read it as testing the generation number itself, not its existence.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -1753,6 +1762,20 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
 }
 
+future<> gossiper::advertise_myself() {
+    if (_initial_generation_number) {
```
like above, I suggest using `_initial_generation_number.has_value()` instead.

tgrabiec commented:
> > > > This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> > > > `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.
> > > 
> > > 
> > > How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.
> > 
> > 
> > When the node restarts, it will indeed inject an entry into the endpoint_state_map (storage_service does this) with generation 0, which is equal to the "shadow" generation used by this patch, so "shadow" gossip will be accepted, not ignored as we want.
> > Secondly, it's not the gossiper component who injects those entries, but scylla's storage service. It's better if gossiper's `advertise_myself::no` is not coupled with external things like that because they can change independently of the gossiper. Suppressing messages within the gossiper doesn't rely on external things like this, so it's better.
> 
> Good that you agree that the !es path is not possible as you claimed earlier. The node will run the remote_generation == local_generation path. However, we do not even need storage service to inject the zero generation to make it work. No node in a working state will have generation zero, such node should never be marked as alive by other nodes. We can skip to mark a node with zero generation to be alive.

I see a contradiction now, and this highlights that there are actually two problems we try to solve here. The contradiction is this: If we're fine with overriding endpoint states on the other nodes, and we only want to prevent the replacing node to be marked alive, then we don't need to gossip with fake generation, we can gossip with the standard generation and only have the replacing node not respond to those requests when it's in the shadow mode.

If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.

I don't know if we want to prevent overriding the state. It seems safer to not override it, but I can't think of any problem with overriding either.

> 
> > > > Not sending responses seems like a safer way to implement a shadow round.
> > > 
> > > 
> > > With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode. It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode.
> > 
> > 
> > The `advertise_myself` mode is also a new mode which can cause regressions. I'm not sure why you think the long shadow round will be less safe. The long shadow round only suppresses messages. Other nodes will perceive it the same way as if the node was partitioned. Sending no messages seems safer than actually sending messages but with a special generation number, because it doesn't affect the state of other nodes.
> 
> No, this patch introduces no new mode. The gossip protocol is not changed. Look at the gossip handlers, nothing has changed there in this patch. No special handling for shadow round. This patch only changes the way the node choose the gossip generation.

Suppressing messages is not exactly changing the protocol, other nodes won't notice anything. Why do you dislike changing the handlers?

> 
> It is not sending any message in most of the cases because other nodes have bigger generation. In special cases when other nodes also think the node has a zero generation, this patch will still work even with this version, because echo message to this replacing node will fail. It is even more safe with v2 below, because other nodes will not try to mark a node with zero generation at all.
> 
> More importantly, this patch fixes an real bug in your proposal - not sending message is not enough.

Yes, but the same fix can be applied to my proposal as well, it doesn't invalidate it.

asias commented:
> > > > > This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> > > > > `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.
> > > > 
> > > > 
> > > > How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.
> > > 
> > > 
> > > When the node restarts, it will indeed inject an entry into the endpoint_state_map (storage_service does this) with generation 0, which is equal to the "shadow" generation used by this patch, so "shadow" gossip will be accepted, not ignored as we want.
> > > Secondly, it's not the gossiper component who injects those entries, but scylla's storage service. It's better if gossiper's `advertise_myself::no` is not coupled with external things like that because they can change independently of the gossiper. Suppressing messages within the gossiper doesn't rely on external things like this, so it's better.
> > 
> > 
> > Good that you agree that the !es path is not possible as you claimed earlier. The node will run the remote_generation == local_generation path. However, we do not even need storage service to inject the zero generation to make it work. No node in a working state will have generation zero, such node should never be marked as alive by other nodes. We can skip to mark a node with zero generation to be alive.
> 
> I see a contradiction now, and this highlights that there are actually two problems we try to solve here. The contradiction is this: If we're fine with overriding endpoint states on the other nodes, and we only want to prevent the replacing node to be marked alive, then we don't need to gossip with fake generation, we can gossip with the standard generation and only have the replacing node not respond to those requests when it's in the shadow mode.

Let me make it clear: 1) the minimum fix is to not responding to the echo message on the replacing node, which fixes on the "receiver" part of this problem. 2) Advertising zero generation is fixing for the "sender" part, so other nodes will not even attempt to mark a node UP in the first place. 1) is a must. 2) is optional and nice to have. 

> 
> If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.
> 
> I don't know if we want to prevent overriding the state. It seems safer to not override it, but I can't think of any problem with overriding either.
> 
> > > > > Not sending responses seems like a safer way to implement a shadow round.
> > > > 
> > > > 
> > > > With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode. It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode.
> > > 
> > > 
> > > The `advertise_myself` mode is also a new mode which can cause regressions. I'm not sure why you think the long shadow round will be less safe. The long shadow round only suppresses messages. Other nodes will perceive it the same way as if the node was partitioned. Sending no messages seems safer than actually sending messages but with a special generation number, because it doesn't affect the state of other nodes.
> > 
> > 
> > No, this patch introduces no new mode. The gossip protocol is not changed. Look at the gossip handlers, nothing has changed there in this patch. No special handling for shadow round. This patch only changes the way the node choose the gossip generation.
> 
> Suppressing messages is not exactly changing the protocol, other nodes won't notice anything. Why do you dislike changing the handlers?

Because it is hard to reason about what is shadow round and long shadow round. The latter is making it even more complicated. One has to reason about which mode the gossip is operating. While we are in the process of getting rid of shadow round, introducing another long shadow round is a big step back.  When GOSSIP_GET_ENDPOINT_STATES is supported on all releases, we can get rid of the old shadow round concept completely. 

> 
> > It is not sending any message in most of the cases because other nodes have bigger generation. In special cases when other nodes also think the node has a zero generation, this patch will still work even with this version, because echo message to this replacing node will fail. It is even more safe with v2 below, because other nodes will not try to mark a node with zero generation at all.
> > More importantly, this patch fixes an real bug in your proposal - not sending message is not enough.
> 
> Yes, but the same fix can be applied to my proposal as well, it doesn't invalidate it.

Of course, anything can be fixed when you know the fix. I am only stating the fact what your proposed in your PR is not a complete solution.

asias commented:
Changes in v3 (6c36dd28f215bd98d7e4784398511a1a9d4b188d):
*) advertise_myself_t to advertise_myself
*) Use _initial_generation_number.has_value()
*) Drop a unused capture for advertise
*) Add comments in gossiper::start_gossiping
*) Drop extra empty line after gossiper::advertise_myself

tgrabiec commented:
On Wed, Jan 13, 2021 at 2:30 AM Asias He <notifications@github.com> wrote:

> This solution introduces a new mechanism in the gossiper which is supposed
> to be the same in its effects as the long shadow round, but rather than
> suppressing message sending it relies on the fact that generation zero will
> be ignored by existing nodes.
> gossiper::do_apply_state_locally() will indeed ignore generations older
> than the one which is already learned. But what if the other node forgot it
> and haven't learnt it yet? Looking at the code, I think it will accept the
> zero generation in the !es path.
>
> How would a node ignore existing node in normal status in
> endpoint_state_map? Even if the node node restarts, it will inject an entry
> in endpoint_state_map.
>
> When the node restarts, it will indeed inject an entry into the
> endpoint_state_map (storage_service does this) with generation 0, which is
> equal to the "shadow" generation used by this patch, so "shadow" gossip
> will be accepted, not ignored as we want.
> Secondly, it's not the gossiper component who injects those entries, but
> scylla's storage service. It's better if gossiper's advertise_myself::no
> is not coupled with external things like that because they can change
> independently of the gossiper. Suppressing messages within the gossiper
> doesn't rely on external things like this, so it's better.
>
> Good that you agree that the !es path is not possible as you claimed
> earlier. The node will run the remote_generation == local_generation path.
> However, we do not even need storage service to inject the zero generation
> to make it work. No node in a working state will have generation zero, such
> node should never be marked as alive by other nodes. We can skip to mark a
> node with zero generation to be alive.
>
> I see a contradiction now, and this highlights that there are actually two
> problems we try to solve here. The contradiction is this: If we're fine
> with overriding endpoint states on the other nodes, and we only want to
> prevent the replacing node to be marked alive, then we don't need to gossip
> with fake generation, we can gossip with the standard generation and only
> have the replacing node not respond to those requests when it's in the
> shadow mode.
>
> Let me make it clear: 1) the minimum fix is to not responding to the echo
> message on the replacing node, which fixes on the "receiver" part of this
> problem. 2) Advertising zero generation is fixing for the "sender" part, so
> other nodes will not even attempt to mark a node UP in the first place. 1)
> is a must. 2) is optional and nice to have.
>

In that case I think we should drop (2) because optimizing this is not
really needed (negligible impact) and we can simplify the code by avoiding
special paths.

I wonder if it's really fine to override the old state. What if the
replacing node fails after overriding the old state. It will show up with
an unknown status in gossip. Can you start another replace operation in
such a state?

If we do want to prevent overriding the old state on other nodes, then
> generation 0 approach is not enough, as it may override the old state in
> some cases.
>
> I don't know if we want to prevent overriding the state. It seems safer to
> not override it, but I can't think of any problem with overriding either.
>
> Not sending responses seems like a safer way to implement a shadow round.
>
> With this patch, we do not need to introduce another gossip mode. It runs
> the normal gossip mode. It is safer in the sense we do not need to worry
> about the regression introduced by this new long shadow round mode.
>
> The advertise_myself mode is also a new mode which can cause regressions.
> I'm not sure why you think the long shadow round will be less safe. The
> long shadow round only suppresses messages. Other nodes will perceive it
> the same way as if the node was partitioned. Sending no messages seems
> safer than actually sending messages but with a special generation number,
> because it doesn't affect the state of other nodes.
>
> No, this patch introduces no new mode. The gossip protocol is not changed.
> Look at the gossip handlers, nothing has changed there in this patch. No
> special handling for shadow round. This patch only changes the way the node
> choose the gossip generation.
>
> Suppressing messages is not exactly changing the protocol, other nodes
> won't notice anything. Why do you dislike changing the handlers?
>
> Because it is hard to reason about what is shadow round and long shadow
> round. The latter is making it even more complicated. One has to reason
> about which mode the gossip is operating.
>
You mean reason form the perspective of gossip internals, if you wanted to
evolve the gossiper code?


> While we are in the process of getting rid of shadow round, introducing
> another long shadow round is a big step back. When
> GOSSIP_GET_ENDPOINT_STATES is supported on all releases, we can get rid of
> the old shadow round concept completely.
>

When GOSSIP_GET_ENDPOINT_STATES is supported, we could get rid of the long
shadow round as well. So it's not making it any harder.


> It is not sending any message in most of the cases because other nodes
> have bigger generation. In special cases when other nodes also think the
> node has a zero generation, this patch will still work even with this
> version, because echo message to this replacing node will fail. It is even
> more safe with v2 below, because other nodes will not try to mark a node
> with zero generation at all.
> More importantly, this patch fixes an real bug in your proposal - not
> sending message is not enough.
>
> Yes, but the same fix can be applied to my proposal as well, it doesn't
> invalidate it.
>
> Of course, anything can be fixed when you know the fix. I am only stating
> the fact what your proposed in your PR is not a complete solution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/7714#issuecomment-759143583>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACFIL4AWMJOHQCJD7M4G5DSZTZT3ANCNFSM4UESTJ5A>
> .
>

asias commented:
> In that case I think we should drop (2) because optimizing this is not really needed (negligible impact) and we can simplify the code by avoiding special paths. I wonder if it's really fine to override the old state. What if the replacing node fails after overriding the old state. It will show up with an unknown status in gossip. Can you start another replace operation in such a state? If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.

Yes, the replacing node can die. It is best to delay the override of the old state as late as possible.  (2) is not just an optimization to avoid some gossip messages over network (which is not important and negligible), but it makes the system more robust in case of the replacing node dies.  BTW, I plan to apply the same method in 829b4c14380020fa4058335c4c43cefec135b3de for replace operation, which depends on (2).  So I am in favour of adds both (1) and (2) now, because I will add (2) anyway. 

Generation zero approach will only override the existing node with generation zero. We can fix this easily by not applying a application state from a remote node with generation zero.

asias commented:
>>   Because it is hard to reason about what is shadow round and long shadow round. The latter is making it even more complicated. One has to reason about which mode the gossip is operating.
>
> You mean reason form the perspective of gossip internals, if you wanted to evolve the gossiper code

Special paths in the gossip handler: if is_in_shadow_round() do this and  if is_in_long_shadow_round() do that

asias commented:
> When GOSSIP_GET_ENDPOINT_STATES is supported, we could get rid of the long shadow round as well. So it's not making it any harder.

GOSSIP_GET_ENDPOINT_STATES is already supported. We just need time to wait for no nodes using the old shadow round methods so we can remove all the special paths  like: 

 302     if (!this->is_enabled() && !this->is_in_shadow_round()) {
 303         return make_ready_future<>();
 304     }


 321         if (this->is_in_shadow_round()) {
 322             this->finish_shadow_round();
 323             // don't bother doing anything else, we have what we came for
 324             return make_ready_future<>();
 325         }

If we introduce the long shadow round, we can not get of the long shadow round even with GOSSIP_GET_ENDPOINT_STATES, because we need it to make replacing working. As a result, we will still have the shadow round concept.

asias commented:
> > In that case I think we should drop (2) because optimizing this is not really needed (negligible impact) and we can simplify the code by avoiding special paths. I wonder if it's really fine to override the old state. What if the replacing node fails after overriding the old state. It will show up with an unknown status in gossip. Can you start another replace operation in such a state? If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.
> 
> Yes, the replacing node can die. It is best to delay the override of the old state as late as possible. (2) is not just an optimization to avoid some gossip messages over network (which is not important and negligible), but it makes the system more robust in case of the replacing node dies. BTW, I plan to apply the same method in [829b4c1](https://github.com/scylladb/scylla/commit/829b4c14380020fa4058335c4c43cefec135b3de) for replace operation, which depends on (2). So I am in favour of adds both (1) and (2) now, because I will add (2) anyway.
> 
> Generation zero approach will only override the existing node with generation zero. We can fix this easily by not applying a application state from a remote node with generation zero.

To be exact, it is not going to override the whole state, it will only apply the state that the local node with generation zero are missing. I see no problem with this.

tgrabiec commented:
> > > In that case I think we should drop (2) because optimizing this is not really needed (negligible impact) and we can simplify the code by avoiding special paths. I wonder if it's really fine to override the old state. What if the replacing node fails after overriding the old state. It will show up with an unknown status in gossip. Can you start another replace operation in such a state? If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.
> > 
> > 
> > Yes, the replacing node can die. It is best to delay the override of the old state as late as possible. (2) is not just an optimization to avoid some gossip messages over network (which is not important and negligible), but it makes the system more robust in case of the replacing node dies. BTW, I plan to apply the same method in [829b4c1](https://github.com/scylladb/scylla/commit/829b4c14380020fa4058335c4c43cefec135b3de) for replace operation, which depends on (2). So I am in favour of adds both (1) and (2) now, because I will add (2) anyway.
> > Generation zero approach will only override the existing node with generation zero. We can fix this easily by not applying a application state from a remote node with generation zero.
> 
> To be exact, it is not going to override the whole state, it will only apply the state that the local node with generation zero are missing. I see no problem with this.

So I suggest this. Since (2) is not needed here, send a version with only (1) as part of this fix. You can introduce (2) where it is really needed. The reason to narrow the changes is reducing risk in case of backports, and also that (2) is better reviewed in the context of the change which really needs it.

tgrabiec commented:
> > When GOSSIP_GET_ENDPOINT_STATES is supported, we could get rid of the long shadow round as well. So it's not making it any harder.
> 
> GOSSIP_GET_ENDPOINT_STATES is already supported. We just need time to wait for no nodes using the old shadow round methods so we can remove all the special paths like:
> 
> 302 if (!this->is_enabled() && !this->is_in_shadow_round()) {
> 303 return make_ready_future<>();
> 304 }
> 
> 321 if (this->is_in_shadow_round()) {
> 322 this->finish_shadow_round();
> 323 // don't bother doing anything else, we have what we came for
> 324 return make_ready_future<>();
> 325 }
> 
> If we introduce the long shadow round, we can not get of the long shadow round even with GOSSIP_GET_ENDPOINT_STATES, because we need it to make replacing working. As a result, we will still have the shadow round concept.

I see what you mean.

Well, the long shadow round will go away after https://github.com/scylladb/scylla/issues/6403

asias commented:
Changes in v4 (0d652f24e0b23442b68a53f75196cba0a5ca5399):
- Change the patch set to contain only the minimum fix to make it easier to backport.

asias commented:
> > > > In that case I think we should drop (2) because optimizing this is not really needed (negligible impact) and we can simplify the code by avoiding special paths. I wonder if it's really fine to override the old state. What if the replacing node fails after overriding the old state. It will show up with an unknown status in gossip. Can you start another replace operation in such a state? If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.
> > > 
> > > 
> > > Yes, the replacing node can die. It is best to delay the override of the old state as late as possible. (2) is not just an optimization to avoid some gossip messages over network (which is not important and negligible), but it makes the system more robust in case of the replacing node dies. BTW, I plan to apply the same method in [829b4c1](https://github.com/scylladb/scylla/commit/829b4c14380020fa4058335c4c43cefec135b3de) for replace operation, which depends on (2). So I am in favour of adds both (1) and (2) now, because I will add (2) anyway.
> > > Generation zero approach will only override the existing node with generation zero. We can fix this easily by not applying a application state from a remote node with generation zero.
> > 
> > 
> > To be exact, it is not going to override the whole state, it will only apply the state that the local node with generation zero are missing. I see no problem with this.
> 
> So I suggest this. Since (2) is not needed here, send a version with only (1) as part of this fix. You can introduce (2) where it is really needed. The reason to narrow the changes is reducing risk in case of backports, and also that (2) is better reviewed in the context of the change which really needs it.

OK. Make sense to me.  I just sent out v4 which contains the minimum changes.

bhalevy commented on gms/gossiper.cc:
```diff
@@ -409,6 +409,9 @@ future<> gossiper::handle_ack2_msg(gossip_digest_ack2 msg) {
 }
 
 future<> gossiper::handle_echo_msg() {
+    if (_initial_generation_number.has_value()) {
```
The name `_initial_generation_number` doesn't make much sense in the context of this minimal patch since we don't do anything with the generation_number value.  We just use the presence of a value to flag this function to throw.

I suggest instead to have a boolean `_advertise_myself` flag a a corresponding method to wrap the logic around it (so the low-level logic can seamlessly change in the future).
The flag can be initialized to `false` and set to `true` in `start_gossiping` if called with `gms::advertise_myself::yes` and in `gossiper::advertise_myself()`

asias replied:
You know I will use _initial_generation_number very soon in the follow up patch. But I do not mind changing it to a bool flag and change it back later.

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1748,6 +1756,13 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
 }
 
+future<> gossiper::advertise_myself() {
+    if (_initial_generation_number.has_value()) {
```
nit: you could drop the if and override unconditionally.

asias replied:
Yes, we could drop it and later bring the check back because we need to do more stuffs inside the if branch.

tgrabiec commented on gms/gossiper.hh:
```diff
@@ -87,6 +87,8 @@ class feature_service;
 struct bind_messaging_port_tag {};
 using bind_messaging_port = bool_class<bind_messaging_port_tag>;
 
+using advertise_myself = bool_class<class advertise_myself_tag>;
```
Please document what this flag means from the perspective of gossiper user who will set the flag.

asias replied:
Done.

asias commented:
Changes in v5 1672f35a2204b59f9d5292c774820b7f43af58c6:
 - Renamed _advertise_myself to _initial_generation_number for now
 - Added doc for the advertise_myself flag
 - Set the flag in gossiper::advertise_myself unconditionally

bhalevy reviewed (approved):
LGTM, thanks

asias commented:
Changes in v6 98999c477ea8f457cccb1f8e36ebdf9c96103302:
- Allow gossiper::handle_echo_msg to run on any shard

tgrabiec commented:
I have a new doubt about the correctness of this solution. I wonder if this race is possible:

1) replacing node sets the status
2) the status propagates to other nodes
3) other nodes send echo
4) echo is ignored by the replacing node
5) replacing node sets _advertise_myself to true
6) replacing node does not update any gossip states so other nodes won't attempt another mark_alive and the node will be considered as not alive

asias commented:
> I have a new doubt about the correctness of this solution. I wonder if this race is possible:
> 
>     1. replacing node sets the status
> 
>     2. the status propagates to other nodes
> 
>     3. other nodes send echo
> 
>     4. echo is ignored by the replacing node
> 
>     5. replacing node sets _advertise_myself to true
> 
>     6. replacing node does not update any gossip states so other nodes won't attempt another mark_alive and the node will be considered as not alive

This is not a concern. The replacing node will update its heartbeat so other nodes will get the update of the endpoint_state (with  bigger hearbeat only but with no application state) and mark replacing node to be alive at line 616 below. 

```
 604         } else if (remote_generation == local_generation) {
 605             if (listener_notification) {
 606                 // find maximum state
 607                 int local_max_version = this->get_max_endpoint_state_version(local_state);
 608                 int remote_max_version = this->get_max_endpoint_state_version(remote_state);
 609                 if (remote_max_version > local_max_version) {
 610                     // apply states, but do not notify since there is no major change
 611                     this->apply_new_states(node, local_state, remote_state);
 612                 } else {
 613                     logger.trace("Ignoring remote version {} <= {} for {}", remote_max_version, local_max_version, node);
 614                 }
 615                 if (!local_state.is_alive() && !this->is_dead_state(local_state)) { // unless of course, it was dead
 616                     this->mark_alive(node, local_state);
 617                 }
```

asias commented:
Changes in v7 1af4b6a5b0f3655d385b48804bc44c2eea265390:
- Limit the gms::advertise_myself::no to replacing_a_node_with_same_ip only.

bhalevy commented on service/storage_service.cc:
```diff
@@ -369,7 +369,8 @@ void storage_service::prepare_to_join(
 
     auto generation_number = db::system_keyspace::increment_and_get_generation().get0();
-    _gossiper.start_gossiping(generation_number, app_states, gms::bind_messaging_port(bool(do_bind))).get();
+    auto advertise = replacing_a_node_with_same_ip ? gms::advertise_myself::no : gms::advertise_myself::yes;
```
I think that `advertise` can be simply constructed as `gms::advertise_myself(!replacing_a_node_with_same_ip)`

bhalevy replied:
The following might work too:
```
gms::advertise_myself advertise = !replacing_a_node_with_same_ip;
```

asias replied:
Yes.

asias commented:
Changed in  v8 c797a561a50e821c6f070343d56d5460c04baca9:
- Use a shorter version to assign advertise

tgrabiec commented on service/storage_service.cc:
```diff
@@ -369,7 +369,8 @@ void storage_service::prepare_to_join(
 
     auto generation_number = db::system_keyspace::increment_and_get_generation().get0();
-    _gossiper.start_gossiping(generation_number, app_states, gms::bind_messaging_port(bool(do_bind))).get();
+    auto advertise = gms::advertise_myself(!replacing_a_node_with_same_ip);
```
Why only in `!replacing_a_node_with_same_ip`, what was the problem ?

asias replied:
1) The problem with replacing using different ip + echo message delay is that, before the existing nodes can mark the replacing node with diff ip address, existing node wont talk to the new replacing node, because it is not in either live nodes list or down node lists. As a result, the replacing node will not learn the gossip state for existing nodes.

2) The issue this patch fixes is limited to replacing_a_node_with_same_ip only. It is best to limit the fix for to it as well.

tgrabiec replied:
I still don't understand. Aren't existing nodes responding to SYNs which the replacing node sends? That's what they do when the node was in the shadow round, and they don't add nodes in the shadow round to live/down lists.

asias replied:
The syn from replacing node contains the node itself only. Gossip only exchanges information for the nodes listed  in the syn message. So the syn from the  replacing node will not make replacing node learn other exiting nodes.

Syn from the existing node to the replacing node (new node) will syn the existing node to the replacing node because existing nodes will contain existing nodes in syn message. However, existing nodes hasn't mark the replacing node as live. So it does not in live list or down list, in each gossip round on existing nodes, no syn message will be send to replacing node. 

Here is an example:

Node 4 is the replacing node replaces Node3 with a diff  ip address. 

![image](https://user-images.githubusercontent.com/320076/105431947-9d0fbe00-5c91-11eb-9da3-ae9347c03646.png)

asias replied:
> Now I understand. So in the case of replacing with the same IP, the node will be seen by existing nodes as down, but they will still gossip with it due to do_gossip_to_unreachable_member()? It gossips with a probability which decreases with the number of live nodes in the cluster. I think we shouldn't rely on it for bootstrap, there's a chance that the node won't be contacted in large clusters. If the SYN didn't contain any digests, the existing node would respond with everything it knows (like in the shadow round). Better do that, that is send a shadow syn when bootstrapping.

The probability on a single node is unreachable_endpoint_count / (live_endpoint_count + 1),  but we have multiple live endpoints in the cluster. Any of them  talks to the replacing node is enough. So the actual probability cluster wide is live_endpoint_count * unreachable_endpoint_count / (live_endpoint_count + 1) which is very big. 

We do not need the trick to empty SYN for the fix to work. It is best to limit the scope of the fix to the minimum. 

BTW: Your reply breaks the thread in the GH page.

tgrabiec replied:
> > Now I understand. So in the case of replacing with the same IP, the node will be seen by existing nodes as down, but they will still gossip with it due to do_gossip_to_unreachable_member()? It gossips with a probability which decreases with the number of live nodes in the cluster. I think we shouldn't rely on it for bootstrap, there's a chance that the node won't be contacted in large clusters. If the SYN didn't contain any digests, the existing node would respond with everything it knows (like in the shadow round). Better do that, that is send a shadow syn when bootstrapping.
> 
> The probability on a single node is unreachable_endpoint_count / (live_endpoint_count + 1), but we have multiple live endpoints in the cluster. Any of them talks to the replacing node is enough. So the actual probability cluster wide is live_endpoint_count * unreachable_endpoint_count / (live_endpoint_count + 1) which is very big.

Your formula can give larger than 1 probability. I think the probability for a single round to contact the node is actually:

`1 - (1 - unreachable_endpoint_count / (live_endpoint_count+1))^(live_endpoint_count)`

Which is about 60% for unreachable_endpoint_count = 1. It quickly approaches 100% after several rounds though, so it's probably still good enough if we wait for several rounds.

Please document the reasons for the special casing (other nodes won't contact us) in the code, and the reasons why we think they will.

asias replied:
Yes,  Pall_not_talk =  (1 - Ptalk) ^ live_endpoint_count is the correct probability. 

Add more docs in v9.

tgrabiec commented:
On Fri, Jan 22, 2021 at 2:11 AM Asias He <notifications@github.com> wrote:

> *@asias* commented on this pull request.
> ------------------------------
>
> In service/storage_service.cc
> <https://github.com/scylladb/scylla/pull/7714#discussion_r562303875>:
>
> > @@ -369,7 +369,8 @@ void storage_service::prepare_to_join(
>      slogger.info("Starting up server gossip");
>
>      auto generation_number = db::system_keyspace::increment_and_get_generation().get0();
> -    _gossiper.start_gossiping(generation_number, app_states, gms::bind_messaging_port(bool(do_bind))).get();
> +    auto advertise = gms::advertise_myself(!replacing_a_node_with_same_ip);
>
> The syn from replacing node contains the node itself only. Gossip only
> exchanges information for the nodes listed in the syn message. So the syn
> from the replacing node will not make replacing node learn other exiting
> nodes.
>
> Syn from the existing node to the replacing node (new node) will syn the
> existing node to the replacing node because existing nodes will contain
> existing nodes in syn message. However, existing nodes hasn't mark the
> replacing node as live. So it does not in live list or down list, in each
> gossip round on existing nodes, no syn message will be sent to replacing
> node.
>

Now I understand.

So in the case of replacing with the same IP, the node will be seen by
existing nodes as down, but they will still gossip with it due
to do_gossip_to_unreachable_member()? It gossips with a probability which
decreases with the number of live nodes in the cluster. I think we
shouldn't rely on it for bootstrap, there's a chance that the node won't be
contacted in large clusters.

If the SYN didn't contain any digests, the existing node would respond with
everything it knows (like in the shadow round). Better do that, that is
send a shadow syn when bootstrapping.

asias commented:
Changes in v9 2e03d9ce2ffba67ffd5e371ea02e204c4852eef4:
- Add more docs explaining the effect of the advertise flag
